File: android/app/src/main/java/com/innocorelabs/verbale/pipeline/DateNorm.kt
package com.innocorelabs.verbale.pipeline

import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.ZoneOffset

/**
 * A spoken date phrase resolved against the meeting's date, or honestly null.
 *
 * A rule, never the model: "Friday" said on a Wednesday is the coming Friday, and that is
 * arithmetic. "Next week", "soon", "another week" have no day in them, and a guess shown as a
 * date is the one thing the record must never do — `date_said` keeps the phrase, `date_norm`
 * stays null, and the item goes to the review queue for a person to pin. Mirrored in
 * src/pipeline/dateNorm.ts; cpp/tests/golden/date_norm.json is the contract between the two.
 */
object DateNorm {
  private val weekdays = mapOf(
    "monday" to DayOfWeek.MONDAY, "tuesday" to DayOfWeek.TUESDAY, "wednesday" to DayOfWeek.WEDNESDAY,
    "thursday" to DayOfWeek.THURSDAY, "friday" to DayOfWeek.FRIDAY, "saturday" to DayOfWeek.SATURDAY,
    "sunday" to DayOfWeek.SUNDAY,
  )
  private val numbers = mapOf(
    "one" to 1, "a" to 1, "two" to 2, "three" to 3, "four" to 4, "five" to 5,
    "six" to 6, "seven" to 7, "eight" to 8, "nine" to 9, "ten" to 10,
  )

  /** The resolved local day as YYYY-MM-DD, or null. The shape the golden table compares. */
  fun resolveDay(said: String, meetingAtMs: Long, zone: ZoneId): String? {
    val today = Instant.ofEpochMilli(meetingAtMs).atZone(zone).toLocalDate()
    return resolveDate(said, today)?.toString()
  }

  /**
   * The resolved calendar day as epoch ms, or null. What `items.date_norm` stores.
   *
   * UTC midnight of the LOCAL calendar day, on purpose: a day is a day, not a moment. Stored as
   * local midnight it would read as the day before to anyone west of the meeting, and the
   * screen's label (recordLabels.ts) reads it back in UTC for the same reason.
   */
  fun resolve(said: String, meetingAtMs: Long, zone: ZoneId): Long? {
    val today = Instant.ofEpochMilli(meetingAtMs).atZone(zone).toLocalDate()
    return resolveDate(said, today)?.atStartOfDay(ZoneOffset.UTC)?.toInstant()?.toEpochMilli()
  }

  /**
   * The date phrase in a statement, as spoken — "Friday", "by the 3rd", "on Thursday",
   * "tomorrow", "in two weeks", "next week" — or null. The first one in the text, with its
   * preposition, in the statement's own casing, so it is verbatim by construction and resolves
   * through [resolveDate] exactly as the model's phrase would. A rule, because a 1.5B model was
   * measured leaving this empty for "Can you send the proposal Friday?"; the model's phrase is
   * only consulted where this finds nothing.
   */
  fun spanIn(text: String): String? = SPAN.find(text)?.value?.trim()?.ifEmpty { null }

  private const val PREP = "(?:\\b(?:by|on|for|until|till|before)\\s+)?"
  private const val WEEKDAY = "(?:monday|tuesday|wednesday|thursday|friday|saturday|sunday)"
  private val SPAN = Regex(
    listOf(
      "$PREP(?:next|this|coming)\\s+$WEEKDAY\\b(?!'s)",
      "$PREP\\b$WEEKDAY\\b(?!'s)",
      "$PREP(?:the\\s+)?\\b\\d{1,2}(?:st|nd|rd|th)\\b",
      "(?:\\bby\\s+)?(?:the\\s+)?\\bend of (?:the )?(?:week|day|month)\\b",
      "\\bin (?:a|an|one|two|three|four|five|six|seven|eight|nine|ten|\\d+) (?:days?|weeks?)\\b",
      "\\bnext (?:week|month)\\b",
      "$PREP\\b(?:today|tonight|tomorrow)\\b",
    ).joinToString("|"),
    RegexOption.IGNORE_CASE,
  )

  fun resolveDate(said: String, today: LocalDate): LocalDate? {
    // Whole phrase, lower case, punctuation gone, leading "by"/"on"/"for" dropped.
    val words = said.lowercase().replace(Regex("[^a-z0-9 ]"), " ").trim().split(Regex("\\s+"))
      .filter { it.isNotEmpty() }
      .let { if (it.isNotEmpty() && it.first() in setOf("by", "on", "for", "until", "till")) it.drop(1) else it }
    if (words.isEmpty()) return null
    val phrase = words.joinToString(" ")

    when (phrase) {
      "today" -> return today
      "tomorrow" -> return today.plusDays(1)
      "end of the week", "end of week", "the end of the week" -> return nextOrToday(today, DayOfWeek.FRIDAY)
    }
    // "next friday": the one after the coming one. "this friday" / "friday": the coming one.
    if (words.size == 2 && words[0] == "next" && words[1] in weekdays) return nextAfterComing(today, weekdays.getValue(words[1]))
    if (words.size == 2 && words[0] == "this" && words[1] in weekdays) return coming(today, weekdays.getValue(words[1]))
    if (words.size == 1 && words[0] in weekdays) return coming(today, weekdays.getValue(words[0]))
    // "the 21st" / "the 3rd": this month, or next if it has passed — and "the 16th" said ON the
    // 16th is next month's: nobody names today's date as a deadline.
    val ordinal = Regex("^(?:the )?([0-9]{1,2})(?:st|nd|rd|th)$").find(phrase)
    if (ordinal != null) {
      val day = ordinal.groupValues[1].toInt()
      if (day in 1..31) {
        val thisMonth = runCatching { today.withDayOfMonth(day) }.getOrNull()
        return if (thisMonth != null && thisMonth.isAfter(today)) thisMonth
        else runCatching { today.plusMonths(1).withDayOfMonth(day) }.getOrNull()
      }
    }
    // "in two weeks" / "in 3 days".
    val inN = Regex("^in ([a-z0-9]+) (day|days|week|weeks)$").find(phrase)
    if (inN != null) {
      val n = inN.groupValues[1].toIntOrNull() ?: numbers[inN.groupValues[1]] ?: return null
      return if (inN.groupValues[2].startsWith("week")) today.plusWeeks(n.toLong()) else today.plusDays(n.toLong())
    }
    return null
  }

  /** The first such weekday strictly after today; the same weekday means a week later. */
  private fun coming(today: LocalDate, day: DayOfWeek): LocalDate {
    var d = today.plusDays(1)
    while (d.dayOfWeek != day) d = d.plusDays(1)
    return d
  }

  private fun nextAfterComing(today: LocalDate, day: DayOfWeek): LocalDate = coming(today, day).plusWeeks(1)

  private fun nextOrToday(today: LocalDate, day: DayOfWeek): LocalDate =
    if (today.dayOfWeek == day) today else coming(today, day)
}

File: src/pipeline/dateNorm.ts
/**
 * A spoken date phrase resolved against the meeting's date, or honestly null.
 *
 * The mirror of DateNorm.kt; cpp/tests/golden/date_norm.json is the contract between the two.
 * The phone writes `date_norm` with the Kotlin one; the screen uses this one to show what a fix
 * would resolve to, and the two must never disagree about which day "Friday" is.
 *
 * All arithmetic is on the meeting's LOCAL calendar day in the given zone, read through Intl so
 * it is right on a phone whose zone is not the meeting's.
 */
const WEEKDAYS: Record<string, number> = {
  sunday: 0, monday: 1, tuesday: 2, wednesday: 3, thursday: 4, friday: 5, saturday: 6,
};
const NUMBERS: Record<string, number> = {
  one: 1, a: 1, two: 2, three: 3, four: 4, five: 5, six: 6, seven: 7, eight: 8, nine: 9, ten: 10,
};
const DROPPED_LEADS = new Set(['by', 'on', 'for', 'until', 'till']);

/** A calendar day, zone-free: the arithmetic happens here and only the ends touch a zone. */
interface Day { y: number; m: number; d: number }

function localDay(atMs: number, timeZone: string): Day {
  const parts = new Intl.DateTimeFormat('en-US', { timeZone, year: 'numeric', month: 'numeric', day: 'numeric' })
    .formatToParts(new Date(atMs));
  const get = (t: string) => Number(parts.find(p => p.type === t)?.value);
  return { y: get('year'), m: get('month'), d: get('day') };
}

const toUtc = (day: Day) => Date.UTC(day.y, day.m - 1, day.d);
const fromUtc = (ms: number): Day => {
  const dt = new Date(ms);
  return { y: dt.getUTCFullYear(), m: dt.getUTCMonth() + 1, d: dt.getUTCDate() };
};
const plusDays = (day: Day, n: number) => fromUtc(toUtc(day) + n * 86_400_000);
const weekdayOf = (day: Day) => new Date(toUtc(day)).getUTCDay();
const isoDay = (day: Day) => `${day.y}-${String(day.m).padStart(2, '0')}-${String(day.d).padStart(2, '0')}`;
const daysInMonth = (y: number, m: number) => new Date(Date.UTC(y, m, 0)).getUTCDate();

/** The first such weekday strictly after today; the same weekday means a week later. */
function coming(today: Day, weekday: number): Day {
  let d = plusDays(today, 1);
  while (weekdayOf(d) !== weekday) d = plusDays(d, 1);
  return d;
}

export function resolveDate(said: string, today: Day): Day | null {
  let words = said.toLowerCase().replace(/[^a-z0-9 ]/g, ' ').trim().split(/\s+/).filter(Boolean);
  if (words.length && DROPPED_LEADS.has(words[0])) words = words.slice(1);
  if (words.length === 0) return null;
  const phrase = words.join(' ');

  if (phrase === 'today') return today;
  if (phrase === 'tomorrow') return plusDays(today, 1);
  if (phrase === 'end of the week' || phrase === 'end of week' || phrase === 'the end of the week') {
    return weekdayOf(today) === WEEKDAYS.friday ? today : coming(today, WEEKDAYS.friday);
  }
  if (words.length === 2 && words[0] === 'next' && words[1] in WEEKDAYS) return plusDays(coming(today, WEEKDAYS[words[1]]), 7);
  if (words.length === 2 && words[0] === 'this' && words[1] in WEEKDAYS) return coming(today, WEEKDAYS[words[1]]);
  if (words.length === 1 && words[0] in WEEKDAYS) return coming(today, WEEKDAYS[words[0]]);

  // "the 21st" / "the 3rd": this month, or next if it has passed — today's own date counts as
  // passed; nobody names today's date as a deadline.
  const ordinal = /^(?:the )?([0-9]{1,2})(?:st|nd|rd|th)$/.exec(phrase);
  if (ordinal) {
    const day = Number(ordinal[1]);
    if (day >= 1 && day <= 31) {
      if (day > today.d && day <= daysInMonth(today.y, today.m)) return { y: today.y, m: today.m, d: day };
      const ny = today.m === 12 ? today.y + 1 : today.y;
      const nm = today.m === 12 ? 1 : today.m + 1;
      return day <= daysInMonth(ny, nm) ? { y: ny, m: nm, d: day } : null;
    }
  }
  const inN = /^in ([a-z0-9]+) (day|days|week|weeks)$/.exec(phrase);
  if (inN) {
    const n = /^[0-9]+$/.test(inN[1]) ? Number(inN[1]) : NUMBERS[inN[1]];
    if (n === undefined) return null;
    return plusDays(today, inN[2].startsWith('week') ? n * 7 : n);
  }
  return null;
}

/** The resolved local day as YYYY-MM-DD, or null. The shape the golden table compares. */
export function resolveDay(said: string, meetingAtMs: number, timeZone: string): string | null {
  const d = resolveDate(said, localDay(meetingAtMs, timeZone));
  return d ? isoDay(d) : null;
}

const WEEKDAY_SHORT = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
const MONTH_SHORT = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

/**
 * A day as the label the queue shows: "Fri 18 Sep". Fixed tables rather than Intl: ICU spells
 * September "Sept" in en-GB on some engines, and a label must read the same on every phone.
 * A number is a stored local midnight, read in UTC (see recordLabels); a string is YYYY-MM-DD.
 */
export function dayLabel(isoOrMs: string | number): string {
  const dt = typeof isoOrMs === 'number' ? new Date(isoOrMs) : new Date(`${isoOrMs}T00:00:00Z`);
  return `${WEEKDAY_SHORT[dt.getUTCDay()]} ${dt.getUTCDate()} ${MONTH_SHORT[dt.getUTCMonth()]}`;
}
